// ts-case-convert/lib/caseConvert.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.toCamel = toCamel;
exports.objectToCamel = objectToCamel;
exports.toSnake = toSnake;
exports.toSnakeNoSplitNumbers = toSnakeNoSplitNumbers;
exports.objectToSnake = objectToSnake;
exports.objectToSnakeNoSplitNumbers = objectToSnakeNoSplitNumbers;
exports.toScreamingSnake = toScreamingSnake;
exports.toScreamingSnakeNoSplitNumbers = toScreamingSnakeNoSplitNumbers;
exports.objectToScreamingSnake = objectToScreamingSnake;
exports.objectToScreamingSnakeNoSplitNumbers = objectToScreamingSnakeNoSplitNumbers;
exports.toPascal = toPascal;
exports.objectToPascal = objectToPascal;
function convertObject(obj, keyConverter) {
    if (obj === null || typeof obj === 'undefined' || typeof obj !== 'object') {
        return obj;
    }
    const out = (Array.isArray(obj) ? [] : {});
    for (const [k, v] of Object.entries(obj)) {
        // @ts-ignore
        out[keyConverter(k)] = Array.isArray(v)
            ? v.map((item) => typeof item === 'object' &&
                !(item instanceof Uint8Array) &&
                !(item instanceof Date)
                ? convertObject(item, keyConverter)
                : item)
            : v instanceof Uint8Array || v instanceof Date
                ? v
                : typeof v === 'object'
                    ? convertObject(v, keyConverter)
                    : v;
    }
    return out;
}
function toCamel(term) {
    if (isNumericKey(term)) {
        return term;
    }
    return (term.length === 1
        ? term.toLowerCase()
        : term
            .replace(/^([A-Z])/, (m) => m[0].toLowerCase())
            .replace(/[_-]([a-z0-9])/g, (m) => m[1].toUpperCase()));
}
function objectToCamel(obj) {
    return convertObject(obj, toCamel);
}
function toSnakeWithSplitNumbers(term, splitNumbers) {
    var _a, _b;
    if (isNumericKey(term)) {
        return term;
    }
    let result = term;
    let circuitBreaker = 0;
    if (splitNumbers) {
        while ((((_a = /([a-z])([0-9]+)(?![0-9]*[A-Z][a-z])/.exec(result)) === null || _a === void 0 ? void 0 : _a.length) || 0) >
            2 &&
            circuitBreaker < 10) {
            result = result.replace(/([a-z])([0-9]+)(?![0-9]*[A-Z][a-z])/, (_all, head, numbers) => `${head.toLowerCase()}_${numbers.toLowerCase()}`);
            circuitBreaker += 1;
        }
    }
    while ((((_b = /(.+?)([A-Z])/.exec(result)) === null || _b === void 0 ? void 0 : _b.length) || 0) > 2 &&
        circuitBreaker < 10) {
        result = result.replace(/(.+?)([A-Z])/, (_all, $1, $2) => `${$1.toLowerCase()}_${$2.toLowerCase()}`);
        circuitBreaker += 1;
    }
    return result.toLowerCase();
}
function toSnake(term) {
    return toSnakeWithSplitNumbers(term, true);
}
function toSnakeNoSplitNumbers(term) {
    return toSnakeWithSplitNumbers(term, false);
}
function objectToSnake(obj) {
    return convertObject(obj, (key) => toSnakeWithSplitNumbers(key, true));
}
function objectToSnakeNoSplitNumbers(obj) {
    return convertObject(obj, (key) => toSnakeWithSplitNumbers(key, false));
}
function toScreamingSnake(term) {
    return toSnake(term).toUpperCase();
}
function toScreamingSnakeNoSplitNumbers(term) {
    return toSnakeNoSplitNumbers(term).toUpperCase();
}
function objectToScreamingSnake(obj) {
    return convertObject(obj, toScreamingSnake);
}
function objectToScreamingSnakeNoSplitNumbers(obj) {
    return convertObject(obj, toScreamingSnakeNoSplitNumbers);
}
function toPascal(term) {
    if (isNumericKey(term)) {
        return term;
    }
    return toCamel(term).replace(/^([a-z])/, (m) => m[0].toUpperCase());
}
function objectToPascal(obj) {
    return convertObject(obj, toPascal);
}
function isNumericKey(term) {
    return /^-?\d+(?:\.\d+)?$/.test(term);
}
